### DISCUSSION TITLE
Design pattern for a one to many where only one of the related objects is typically used

### DISCUSSION THREAD

**Discussion by aathan** [sqlalchemy/sqlalchemy#7585] [Usage Questions]
Could someone point me to this usecase's recommended design pattern in SQLA? I have reviewed https://docs.sqlalchemy.org/en/14/orm/collections.html but may have missed the relevant case. Also, I don't believe AssociationProxy is relevant (but I did submit a pull request with some doc clarifications).

ClassA has a one to many relationship ClassA.current_b to ClassB. When we perform operations on ClassA we may create additional ClassB's, but we really only care to reference/access the most recent ClassB. We may or may not want to load all ClassB's but we really only care about the "most recent" one. 

An obvious solution is to have an all_bs relationship with a sorted container plus an @property that fetches all_bs[-1]. 

More generally, there is a collection of ClassB's but we only care to access some reduction on that collection (e.g., min(ClassA.all_bs) etc).

Obviously, this can get complex, but in this example and possibly others, the reduction is expressible as SQL (e.g., as a subquery on class_b to fetch the most recent one). Perhaps there is a feature on relationship that allows the fetch to be for some subset of the related-to objects?

There are many degrees of freedom on possible solutions, but I'm wondering what the SQLA experts think about this.

**Comment by zzzeek** [comment]
if there is no additional foreign key column pointing to "the B we care about" and you want to use SQL functions to detemine the "B we care about", the best pattern we have uses window functions and is illustrated at https://docs.sqlalchemy.org/en/14/orm/join_conditions.html#row-limited-relationships-with-window-functions.  There are other ways to do this pattern without window functions as well but they typically require correlated subqueries, so if you are on a modern database engine, this recipe should work well.

OTOH if you have a column on A that refers to the single B in the collection you care about, meaning you have foreign key columns pointing in both directions, that pattern is suited by the "post_update" persistence feature and there's an example of that at https://docs.sqlalchemy.org/en/14/orm/relationship_persistence.html#rows-that-point-to-themselves-mutually-dependent-rows .

**Comment by aathan** [comment]
Since posting my question, I also found

https://docs.sqlalchemy.org/en/14/orm/join_conditions.html#specifying-alternate-join-conditions

which seem relevant. Here, it would seem possible to create a join condition that picks out only one of the related-to objects on some arbitrary condition. I imagine then, that if there is no cascade delete on the relationship, I could change the .current_b attribute of ClassA to some new ClassB instance, and the database would have all the prior ClassB's, while the in-memory ClassA would point only to the new one AND fetches would use the foreign() or remote() annotations to fetch only one of them?

A possible twist here is that I don't want a collection in the attribute value, just "the one object."

Am I remotely (bada-ching!) pointed in the right direction?

**Comment by aathan** [comment]
I guess I'm more specifically looking for a worked example where the join-fetched relationship uses an "ordered by" on the related to object and loads only the "limit 1" top result. I have to make progress so I'm going to kludge something for now and try to play with the APIs later, but "oye."
